### PR TITLE
EXSWHTEC-97 - Implement tests for hipMemcpy2D APIs

### PR DESCRIPTION
**Pull request by mirza-halilcevic**
- Implement basic behavior checks in all copy directions
- Implement synchronization behavior checks for expected behavior based on cuda docs
- Implement positive tests for zero sized width and/or height copies, where no copy is expected to happen
- Implement negative parameter tests
- Implement all of the above for hipMemcpy2D and hipMemcpy2DAsync.

### CODE REVIEW

**Comment by jaydeeppatel1111** [comment]
Memcpy2DHtoHSyncBehavior with malloc case is fixed and will be available in upcoming release. however We should use async version of API to get async behavior for DD. i.e. Memcpy2DDtoDSyncBehavior can be fixed through test case change. Currently tests use sync version of API for DD that can be changed to Async version. True, Which is perhaps not matching to cuda however we have decided to keep it as is. Please update test cases and let us know if there is any issue. Thanks
@chrispaquot @mangupta

**Comment by mirza-halilcevic** [comment]
> Memcpy2DHtoHSyncBehavior with malloc case is fixed and will be available in upcoming release. however We should use async version of API to get async behavior for DD. i.e. Memcpy2DDtoDSyncBehavior can be fixed through test case change. Currently tests use sync version of API for DD that can be changed to Async version. True, Which is perhaps not matching to cuda however we have decided to keep it as is. Please update test cases and let us know if there is any issue. Thanks @chrispaquot @mangupta

The test case is now updated.

File: catch/unit/memory/hipMemcpy2D.cc [hunk 1]
```diff
@@ -1,284 +1,151 @@
...
+  SECTION("Device to Host") {
+    Memcpy2DDtoHPageableSyncBehavior(hipMemcpy2D, true);
+    Memcpy2DDtoHPinnedSyncBehavior(hipMemcpy2D, true);
+  }
 
-      // Host to Device
-      HIP_CHECK(hipMemcpy2D(A_d, pitch_A, A_h, COLUMNS*sizeof(TestType),
-            COLUMNS*sizeof(TestType), ROWS, hipMemcpyHostToDevice));
+  SECTION("Device to Device") {
+#if HT_NVIDIA
```
Comment (chrispaquot) [inline catch/unit/memory/hipMemcpy2D.cc:56]: I think we fixed this, no?

Comment (mirza-halilcevic) [reply catch/unit/memory/hipMemcpy2D.cc:56]: As per request on the related issue (EXSWHTEC-232), the test case is changed to expect asynchronous behavior on NVIDIA, and synchronous behavior on AMD in this case.

**Review by chrispaquot — APPROVED** [review]

**Comment by rorake** [comment]
@mirza-halilcevic 
Will you please resolve the merge conflicts for this PR?

**Comment by rakesroy** [comment]
PR has been merged into develop branch via commit c9e664b8e646404731b40cfc2e4d38f07a62e3aa.

**Pull request closed without merging**